### PR TITLE
Add query to check if npm install is executed without versioning. Closes #1042

### LINKED ISSUE #1042: Add NPM Install Command Without Pinned Version query for Docker
### Platform
Docker

### Description
Check if packages installed by npm are pinning a specific version.

### PR DESCRIPTION
**Pull request by joaomartinscx**
Check if any npm install execution sets the version of the package. Closes #1042

### CODE REVIEW

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+  atIndex != -1 #package must refer the version or tag
+} else {
+  hasScope := regex.match(install,"@.+\/")
+  not hasScope
+  atIndex := indexof(install,"@")
+  atIndex != -1 #package must refer the version or tag
+} else = false
+
+getNPMInstallCmdIdx(value) = idx {
+  some i
```
Comment (joaoReigota1) [inline assets/queries/dockerfile/npm_install_without_pinned_version/query.rego:85]: 'some i' not needed here

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+CxPolicy [ result ] {
+  runCmd := input.document[i].command[name][_]
+  isRunCmd(runCmd) 
+  
+  value := runCmd.Value
+  count(value) == 1 #command is in a single string
+
+  cmd := value[0]
+
+  searchIndex := indexof(cmd,"npm install")
```
Comment (joaoReigota1) [inline assets/queries/dockerfile/npm_install_without_pinned_version/query.rego:12]: Should check for 'npm i' and 'npm add' as well

**Review by joaoReigota1 — CHANGES_REQUESTED** [review]

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego (hunk 2 above)
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+                "searchKey": 	    sprintf("FROM={{%s}}.{{%s}}", [name,runCmd.Original]),
+                "issueType":		"IncorrectValue", 
+                "keyExpectedValue": sprintf("'%s' uses npm install with a pinned version", [runCmd.Original]),
+                "keyActualValue": 	sprintf("'%s' does not uses npm install with a pinned version", [runCmd.Original])
+              }
+}
+
+isRunCmd(com) = true{
+  com.Cmd == "run"
+} else = false
```
[the remark this answers is not in the document]
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+
+trimCmdEnd(cmd) = trimmed {
+  termOps := ["&&","||","|","&",";"]
+
+  splitStr := split(cmd," ")
+  some i, j
```
[the remark this answers is not in the document]
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+  scopeEnd := indexof(install,"/")
+  packageID := substring(install,scopeEnd+1,count(install) - scopeEnd)
+  atIndex := indexof(packageID,"@")
+  atIndex != -1 #package must refer the version or tag
+} else {
+  hasScope := regex.match(install,"@.+\/")
+  not hasScope
+  atIndex := indexof(install,"@")
+  atIndex != -1 #package must refer the version or tag
+} else = false
```
[the remark this answers is not in the document]
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego (hunk 1 above)
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

File: assets/queries/dockerfile/npm_install_without_pinned_version/query.rego [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+
+getNPMInstallCmdIdx(value) = idx {
+  some i
+    value[i] == "npm"
+    value[i+1] == "install"
+    idx := i+2
+} else = -1
```
[the remark this answers is not in the document]
Comment (joaomartinscx) [reply assets/queries/dockerfile/npm_install_without_pinned_version/query.rego]: fixed

**Review by joaomartinscx — COMMENTED** [review]
Added requested changes

**Review by joaoReigota1 — APPROVED** [review]

**Pull request merged**